Introduction
------------

Physher is a maximum likelihood-based method for estimating evolutionary rates and divergence times from genetic, amino acid, codon, and generic data. Phylogenetic trees should be open with Figtree from this  [website](http://tree.bio.ed.ac.uk/software/figtree/) or [Seqotron] (https://github.com/4ment/seqotron/).

For more details about the configuration file and command line arguments see the [wiki](https://github.com/4ment/physher/wiki/Usage)


Features
--------

 * Substitution models:
  * Nucleotide: Reversible: GTR, HKY, K80, and JC69 + any models using encoding 00000 ... 01234. Non reversible:  UREV (stationary) and NONSTAT (non stationary).
  * Codon: GY94.
  * Amino acid: Dayhoff, LG, and WAG.
  * General: custom state space (e.g. phylogeography).

 * Rate heterogeneity among sites:
  * Gamma distributed.
  * Proportion of invariant site.

 * Molecular clock:
  * No clock
  * Strict clock: constant rate across lineages.
  * Local clock: monophyletic lineages evolve at the same rate.
  * Discrete clock: a discrete distribution of rates is assigned to lineages without any restriction on the rate distribution.
 
 * Search algorithms for clock models:
  * Greedy algorithm (local clock only).
  * Genetic algorithm (discrete clock only).
 
 * Topology search:
  * Nearest neighbor interchange (NNI)

 * Confidence intervals
  * Non-parametric bootstrap (Normal and percentile) using _bootstrap_ program (only available for strict clock models).

 * Model averaging of divergence times and substitution rate
  * Using modelavg program (only available using the genetic algorithm with discrete clocks).


Getting started
---------------

[Install physher](https://github.com/4ment/physher/wiki/Install)

[Online manual](https://github.com/4ment/physher/wiki/Usage)


Associated programs
-------------------------

 * _bootstrap_: calculate confidence intervals using bootstrap trees generated by a strict clock with _physher_. Every tree need to have the same topology.
 * _simultron_: simulate nucleotide data given a fixed tree topology.
 * _modelavg_: model average of trees generated by genetic algorithm with a discrete clock.


Reference
---------

Fourment M and Holmes EC. Novel non-parametric models to estimate evolutionary rates and divergence times from heterochronous sequence data. BMC Evolutionary Biology 14:163, 2014.